> One last comment, and then I think there's a style issue, which can be fixed and this is good to merge. If you run `black .` and `isort .`, or install the pre-commit hooks for networkx, it should fix the errors.
> 
> https://github.com/networkx/networkx/blob/main/CONTRIBUTING.rst
> 
> ```
> black....................................................................Failed
> [29](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:30)
> - hook id: black
> [30](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:31)
> - files were modified by this hook
> [31](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:32)
> 
> [32](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:33)
> reformatted networkx/algorithms/causal/m_separation.py
> [33](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:34)
> reformatted networkx/algorithms/causal/tests/test_m_separation.py
> [34](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:35)
> 
> [35](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:36)
> All done! ✨ 🍰 ✨
> [36](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:37)
> 2 files reformatted, 620 files left unchanged.
> [37](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:38)
> 
> [38](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:39)
> pyupgrade................................................................Passed
> [39](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:40)
> blacken-docs.............................................................Passed
> [40](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:41)
> isort (python)...........................................................Failed
> [41](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:42)
> - hook id: isort
> [42](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:43)
> - files were modified by this hook
> [43](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:44)
> 
> [44](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:45)
> Fixing /home/runner/work/networkx/networkx/networkx/algorithms/causal/m_separation.py
> [45](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:46)
> Fixing /home/runner/work/networkx/networkx/networkx/algorithms/causal/tests/test_m_separation.py
> [46](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:47)
> Skipped 50 files
> [47](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:48)
> 
> [48](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:49)
> prettier.................................................................Passed
> [49](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:50)
> pre-commit hook(s) made changes.
> [50](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:51)
> If you are seeing this message in CI, reproduce locally with: `pre-commit run --all-files`.
> [51](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:52)
> To run `pre-commit` as part of git workflow, use `pre-commit install`.
> [52](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:53)
> All changes made by hooks:
> [53](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:54)
> diff --git a/networkx/algorithms/causal/m_separation.py b/networkx/algorithms/causal/m_separation.py
> [54](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:55)
> index 2dee20b..cdca12f 100644
> [55](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:56)
> --- a/networkx/algorithms/causal/m_separation.py
> [56](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:57)
> +++ b/networkx/algorithms/causal/m_separation.py
> [57](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:58)
> @@ -1,10 +1,9 @@
> [58](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:59)
> -from copy import deepcopy
> [59](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:60)
>  from collections import deque
> [60](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:61)
> +from copy import deepcopy
> [61](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:62)
>  
> [62](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:63)
>  import networkx as nx
> [63](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:64)
>  from networkx.utils import UnionFind
> [64](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:65)
>  
> [65](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:66)
> -
> [66](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:67)
>  __all__ = ["m_separated"]
> [67](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:68)
>  
> [68](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:69)
>  
> [69](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:70)
> @@ -80,16 +79,12 @@ def m_separated(
> [70](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:71)
>      # get bidirected edges subgraph
> [71](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:72)
>      G_bidirected = nx.Graph()
> [72](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:73)
>      G_bidirected.add_nodes_from((n, deepcopy(d)) for n, d in G.nodes.items())
> [73](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:74)
> -    G_bidirected.add_edges_from(
> [74](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:75)
> -        G.get_graphs(edge_type=bidirected_edge_name).edges
> [75](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:76)
> -    )
> [76](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:77)
> +    G_bidirected.add_edges_from(G.get_graphs(edge_type=bidirected_edge_name).edges)
> [77](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:78)
>  
> [78](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:79)
>      # get ancestral subgraph of x | y | z by removing leaves in directed graph that are not in x | y | z
> [79](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:80)
>  
> [80](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:81)
>      # until no more leaves can be removed.
> [81](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:82)
> -    leaves = deque(
> [82](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:83)
> -        [n for n in G_directed.nodes if G_directed.out_degree[n] == 0]
> [83](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:84)
> -    )
> [84](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:85)
> +    leaves = deque([n for n in G_directed.nodes if G_directed.out_degree[n] == 0])
> [85](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:86)
>      while len(leaves) > 0:
> [86](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:87)
>          leaf = leaves.popleft()
> [87](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:88)
>          if leaf not in union_xyz:
> [88](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:89)
> diff --git a/networkx/algorithms/causal/tests/test_m_separation.py b/networkx/algorithms/causal/tests/test_m_separation.py
> [89](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:90)
> index 0bda958..58faafc 100644
> [90](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:91)
> --- a/networkx/algorithms/causal/tests/test_m_separation.py
> [91](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:92)
> +++ b/networkx/algorithms/causal/tests/test_m_separation.py
> [92](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:93)
> @@ -1,6 +1,7 @@
> [93](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:94)
> -import networkx as nx
> [94](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:95)
>  import pytest
> [95](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:96)
>  
> [96](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:97)
> +import networkx as nx
> [97](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:98)
> +
> [98](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:99)
>  
> [99](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:100)
>  def test_m_separation():
> [100](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:101)
>      digraph = nx.path_graph(4)
> [101](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:102)
> @@ -10,7 +11,9 @@ def test_m_separation():
> [102](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:103)
>      G = nx.MixedEdgeGraph([digraph, bigraph], ["directed", "bidirected"])
> [103](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:104)
>  
> [104](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:105)
>      # error should be raised if someone does not use a MixedEdgeGraph
> [105](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:106)
> -    with pytest.raises(RuntimeError, match="m-separation should only be run on a MixedEdgeGraph"):
> [106](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:107)
> +    with pytest.raises(
> [107](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:108)
> +        RuntimeError, match="m-separation should only be run on a MixedEdgeGraph"
> [108](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:109)
> +    ):
> [109](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:110)
>          nx.m_separated(digraph, {0}, {1}, set())
> [110](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:111)
>  
> [111](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:112)
>      # basic d-separation statements based on blocking paths should work
> [112](https://github.com/adam2392/networkx/runs/8114578803?check_suite_focus=true#step:5:113)
> Error: Process completed with exit code 1.
> ```

Yep I will get onto those pre-commit hooks

--- a/networkx/algorithms/causal/__init__.py
+++ b/networkx/algorithms/causal/__init__.py
@@ -1,2 +1,2 @@
-from .m_separation import *  # noqa: F403
 from .convert import *  # noqa: F403
+from .m_separation import *  # noqa: F403

--- a/networkx/algorithms/causal/m_separation.py
+++ b/networkx/algorithms/causal/m_separation.py
@@ -1,4 +1,8 @@
+from collections import deque
+from copy import deepcopy
+
 import networkx as nx
+from networkx.utils import UnionFind
 
 __all__ = ["m_separated"]
 
@@ -8,14 +12,14 @@ def m_separated(
 ):
     """Check m-separation among 'x' and 'y' given 'z' in mixed-edge causal graph G.
 
-    This algorithm wraps ``networkx.algorithms.d_separated``, but
-    allows one to pass in a ``ADMG`` instance instead.
+    This algorithm adapts the linear time algorithm presented in [1]_ currently implemeted in
+    networkx.algorithms.d_separation to work for mixed-edge causal graphs.
 
-    It first converts all bidirected edges into explicit unobserved
-    confounding nodes in an explicit ``networkx.DiGraph``, which then
-    calls ``networkx.algorithms.d_separated`` to determine d-separation.
-    This inherently increases the runtime cost if there are many
-    bidirected edges, because many nodes must be added.
+    This algorithm first obtains the ancestral subgraph of x | y | z which only requires knowledge of
+    the directed edges. Then, all outgoing directed edges from nodes in z are deleted. After that, an
+    undirected graph composed from the directed and bidirected edges amongst the remaining nodes is
+    created. Then, x is independent of y given z if x is disconnected from y in this new graph.
+
 
     Parameters
     ----------
@@ -32,6 +36,13 @@ def m_separated(
     -------
     b : bool
         A boolean that is true if ``x`` is m-separated from ``y`` given ``z`` in ``G``.
+
+
+    References
+    ----------
+
+    .. [1] Darwiche, A.  (2009).  Modeling and reasoning with Bayesian networks.
+       Cambridge: Cambridge University Press.
 
     See Also
     --------
@@ -52,27 +63,57 @@ def m_separated(
         for edge_type in [bidirected_edge_name, directed_edge_name]
     ):
         raise RuntimeError(
-            f"m-separation only works on nx with directed and bidirected edges. "
+            f"m-separation only works on graphs with directed and bidirected edges. "
             f"Your graph passed in has the following edge types: {G.edge_types}, whereas "
             f"the function is expecting directed edges named {directed_edge_name} and "
             f"bidirected edges named {bidirected_edge_name}."
         )
 
-    # get the full graph by converting bidirected edges into latent confounders
-    # and keeping the directed edges
-    explicit_G = nx.bidirected_to_unobserved_confounder(
-        G, bidirected_edge_name=bidirected_edge_name
-    )
+    if not nx.is_directed_acyclic_graph(G.get_graphs(directed_edge_name)):
+        raise nx.NetworkXError("directed edge graph should be directed acyclic")
 
-    # Convert the graph to a directed graph;
-    # At this point, it should be a valid DAG if the original graph was acyclic
-    explicit_G = explicit_G.to_directed()
+    union_xyz = x.union(y).union(z)
 
-    # get all unobserved confounders
-    uc_nodes = {
-        node for node, label in explicit_G.nodes(data="label") if label is not None
-    }
+    # get directed edges
+    G_directed = nx.DiGraph()
+    G_directed.add_nodes_from((n, deepcopy(d)) for n, d in G.nodes.items())
+    G_directed.add_edges_from(G.get_graphs(edge_type=directed_edge_name).edges)
 
-    # make sure there are always conditioned on the conditioning set
-    assert all(uc_node not in z for uc_node in uc_nodes)
-    return nx.d_separated(explicit_G, x, y, z)
+    # get bidirected edges subgraph
+    G_bidirected = nx.Graph()
+    G_bidirected.add_nodes_from((n, deepcopy(d)) for n, d in G.nodes.items())
+    G_bidirected.add_edges_from(G.get_graphs(edge_type=bidirected_edge_name).edges)
+
+    # get ancestral subgraph of x | y | z by removing leaves in directed graph that are not in x | y | z
+
+    # until no more leaves can be removed.
+    leaves = deque([n for n in G_directed.nodes if G_directed.out_degree[n] == 0])
+    while len(leaves) > 0:
+        leaf = leaves.popleft()
+        if leaf not in union_xyz:
+            for p in G_directed.predecessors(leaf):
+                if G_directed.out_degree[p] == 1:
+                    leaves.append(p)
+            G_directed.remove_node(leaf)
+            G_bidirected.remove_node(leaf)
+
+    # remove outgoing directed edges in z
+    edges_to_remove = list(G_directed.out_edges(z))
+    G_directed.remove_edges_from(edges_to_remove)
+
+    # make new undirected graph from remaining directed and bidirected edges
+    G_final = nx.Graph()
+    G_final.add_nodes_from((n, deepcopy(d)) for n, d in G_final.nodes.items())
+    G_final.add_edges_from(G_directed.edges)
+    G_final.add_edges_from(G_bidirected.edges)
+
+    disjoint_set = UnionFind(G_final.nodes())
+    for component in nx.connected_components(G_final):
+        disjoint_set.union(*component)
+    disjoint_set.union(*x)
+    disjoint_set.union(*y)
+
+    if x and y and disjoint_set[next(iter(x))] == disjoint_set[next(iter(y))]:
+        return False
+    else:
+        return True

--- a/networkx/algorithms/causal/tests/test_m_separation.py
+++ b/networkx/algorithms/causal/tests/test_m_separation.py
@@ -1,5 +1,6 @@
-import networkx as nx
 import pytest
+
+import networkx as nx
 
 
 def test_m_separation():
@@ -10,7 +11,9 @@ def test_m_separation():
     G = nx.MixedEdgeGraph([digraph, bigraph], ["directed", "bidirected"])
 
     # error should be raised if someone does not use a MixedEdgeGraph
-    with pytest.raises(RuntimeError, match="m-separation should only be run on a MixedEdgeGraph"):
+    with pytest.raises(
+        RuntimeError, match="m-separation should only be run on a MixedEdgeGraph"
+    ):
         nx.m_separated(digraph, {0}, {1}, set())
 
     # basic d-separation statements based on blocking paths should work

--- a/networkx/algorithms/causal/tests/test_m_separation.py
+++ b/networkx/algorithms/causal/tests/test_m_separation.py
@@ -1,21 +1,18 @@
-import pytest
-
 import networkx as nx
+import pytest
 
 
 def test_m_separation():
     digraph = nx.path_graph(4)
+    digraph.add_edge(2, 4)
     bigraph = nx.Graph([(2, 3)])
     bigraph.add_nodes_from(digraph)
     G = nx.MixedEdgeGraph([digraph, bigraph], ["directed", "bidirected"])
 
     # error should be raised if someone does not use a MixedEdgeGraph
-    with pytest.raises(
-        RuntimeError, match="m-separation should only be run on a MixedEdgeGraph"
-    ):
+    with pytest.raises(RuntimeError, match="m-separation should only be run on a MixedEdgeGraph"):
         nx.m_separated(digraph, {0}, {1}, set())
 
-    print(G.edges())
     # basic d-separation statements based on blocking paths should work
     assert not nx.m_separated(G, {0}, {3}, set())
     assert nx.m_separated(G, {0}, {3}, {1})
@@ -23,3 +20,6 @@ def test_m_separation():
     # conditioning on a collider via bidirected edge opens the collider
     assert not nx.m_separated(G, {0}, {3}, {2})
     assert nx.m_separated(G, {0}, {3}, {1, 2})
+
+    # conditioning on a descendant of a collider via bidirected edge opens the collider
+    assert not nx.m_separated(G, {0}, {3}, {4})

--- a/networkx/classes/mixededge.py
+++ b/networkx/classes/mixededge.py
@@ -1,6 +1,7 @@
 from copy import deepcopy
 from functools import cached_property
 from itertools import chain
+from typing import Iterable, Iterator
 
 import networkx as nx
 from networkx.classes.reportviews import NodeView
@@ -1054,10 +1055,11 @@ class MixedEdgeGraph:
             for edge_type_, graph in self._edge_graphs.items()
         }
 
-    def neighbors(self, n):
-        return chain.from_iterable(
-            nx.all_neighbors(G, n) for _, G in self.get_graphs().items()
-        )
+    def neighbors(self, n) -> Iterator:
+        nbrs = set()
+        for _, G in self.get_graphs().items():
+            nbrs = nbrs.union(set(nx.all_neighbors(G, n)))
+        return iter(nbrs)
 
     def subgraph(self, nodes):
         """Returns a SubGraph view of the subgraph induced on ``nodes``.

